# Exercise 1

## Step 1.1

In [848]:
cards = [1,2,3,4,5,6,7,8,9,10,11,12,13]*4
print("Standard deck of playing cards will consist of :"+ str(cards))
print("where 1=Ace, 11=Jack, 12=Queen, 13=King that is total of " + str(len(cards)) + " cards")

Standard deck of playing cards will consist of :[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
where 1=Ace, 11=Jack, 12=Queen, 13=King that is total of 52 cards


## Step 1.2

In [844]:
total = sum(cards)
print("Total value associated with full deck which represents maximum total for a game is" + str(total))

Total value associated with full deck which represents maximum total for a game is364


# Exercise 2

## Step 2.1

In [534]:
#definig deal function here where cards list is created with 52 cards. 
from random import shuffle    #importing shuffle function to use it for shuffling the cards.
def deal():
    cards = [1,2,3,4,5,6,7,8,9,10,11,12,13]*4
    player_1=()
    player_2=()
    shuffle(cards)            #this function changes the original cards list.  
    cards1=cards
    n=len(cards)
    index = int(n/2)
    
    player_1=cards[:index]    #gives first 26 cards to player 1
    player_2=cards[index:n]   #gives last 26 cards to player 2 
    return(tuple(player_1),tuple(player_2))

## Step 2.2

In [849]:
a,b=deal() #using deal function to give one hand to Player1 that is a and one hand to Player2 that is b

print("Player 1's hand: ")
print(a)
print("Player 1's number of cards: ")
print(len(a))
print("Player 1's total of cards: ")
print(sum(a))

print("Player 2's hand: ")
print(b)
print("Player 2's number of cards: ")
print(len(b))
print("Player 2's total of cards: ")
print(sum(b))

Player 1's hand: 
(3, 13, 3, 9, 9, 9, 12, 4, 6, 12, 8, 11, 11, 8, 8, 2, 5, 4, 5, 1, 6, 4, 13, 5, 10, 6)
Player 1's number of cards: 
26
Player 1's total of cards: 
187
Player 2's hand: 
(3, 9, 5, 7, 2, 13, 7, 8, 7, 10, 1, 12, 2, 4, 11, 10, 12, 10, 11, 1, 3, 13, 6, 1, 2, 7)
Player 2's number of cards: 
26
Player 2's total of cards: 
177


# Excercise 3 

## Step 3.1

#### Steps of what should occur in game of addition war and process behind play() function:

1. Shuffle Deck - which is done by deal() function.

2. Deal half of deck to each player - deal() function.

3. Empty discard piles for P1 and P2 are created which will store the cards of the round for final calculation if a Player wins that round.

4. i is used to keep check of while loop index which infact is also keeping track of element selection from a and b which are 2 hands of players.

5. using a[i] and a[i+1], two consecutive cards are used to play addition war.

6. After entering the loop, if there is tie, cards are sent over to tiebreaker function which takes in a,b,i and inc where inc increment value of i for outer while loop. example, if elements 1,2 are tie for a and elements 3,4 are used to break the tie, then elements 5,6 should be used in next round after tiebreaker, so inc gives appropriate value to increment while after tiebreaker.

7. In tiebreaker, if next cards are giving a tie result again, then a recursive call to tiebreaker function is made otherwise winner is sent back to the function call.

8. In tiebreaker, Exhausted takes care of situation where the last 2 cards can go in tiebreaker and there are no cards to decide the result.

9. Appropriate cards are added to discard piles of Player1 and Player2 with help of return values from tiebreaker function.

10. Using inc value from tiebrekaer, increment is made in such way that no cards are repeated in next round after tiebrekaer.

11. If the sum of cards is not equal then there is no calling for tiebreaker and normal if else conditions are used to compare the cards addition. At the end, winner gets all cards of that round in his/her discard pile, including opponent's.

12. i value for while loop is incremented at end of every loop.

13. After each round is played, comparison of discard piles is done and values are returned according to situations which are if Player1 wins or Player2 wins or there is a tie.

14. The return values are used to print which is winner with total of discard pile and which is loser with total of discard pile.

15. Game terminates after maximum of 13 rounds if there is no tie. If there is one tie, there are 12 rounds, and so on.

16. The player with the highest total card points (in their discard pile) wins the game.

## Step 3.2

In [880]:
def play():    
    a,b=deal()
    
    discard_pile_p1 = [] #player1 discard pile
    discard_pile_p2 = [] #player2 discard pile
    
    i=0

    while i<=25:

        if(i%2==0):  #using this to fiter 2 cards together and not repeat the cards in next round

            if((a[i]+a[i+1])==(b[i]+b[i+1])):  #condition where a tie can occur
                
                tie_pile = []; inc = 0; player_won = ""; p1_exhausted = []; p2_exhausted = [];
                tie_pile,inc,player_won = tiebreaker(a,b,i,inc) #call to tiebreaker function
                
                if player_won == "Player 1":
                    discard_pile_p1.extend(tie_pile)
                    print("Tie first, but Player1 won")
                    print("")

                elif player_won == "Player 2":
                    discard_pile_p2.extend(tie_pile)
                    print("Tie first, but Player2 won")
                    print("")
                    
                elif player_won == "Exhausted": 
                    print("Tie, and cards exhausted")
                    print("")
                    
                    p1_exhausted = [a[i],a[i+1]]
                    discard_pile_p1.extend(p1_exhausted)
                    
                    p2_exhausted = [b[i],b[i+1]]
                    discard_pile_p2.extend(p2_exhausted)
                    
                i += inc - 1   #increment value of i loop according to number of ties occured so as to skip cards which were used to break the tie.
                
                
            elif((a[i]+a[i+1])>(b[i]+b[i+1])):   #condition where player1 wins
                print("Player1 wins")
                print("")
                discard_pile_p1.extend([a[i],a[i+1],b[i],b[i+1]])
                

            elif((a[i]+a[i+1])<(b[i]+b[i+1])):  #condition where player2 wins
                print("Player2 wins")
                print("")
                discard_pile_p2.extend([a[i],a[i+1],b[i],b[i+1]])
                    
                
        i += 1    #incrementing the while loop   
    
   
    
    if(sum(discard_pile_p1)>sum(discard_pile_p2)): #returning values if player1's discard_pile is gretaer, that is Player1 wins
        return sum(discard_pile_p1),sum(discard_pile_p2),"Player 1","Player 2"
    
    elif(sum(discard_pile_p1)<sum(discard_pile_p2)): ##returning values if player2's discard_pile is gretaer, that is Player2 wins
        return sum(discard_pile_p2),sum(discard_pile_p1),"Player 2","Player 1"
    
    else: #returning values if there is a tie in total value of final piles.
        return sum(discard_pile_p1),sum(discard_pile_p2),"Tie","Tie"
    
    print("")
    
        
    
def tiebreaker(a,b,i,inc): #tiebreaker function to give result for tie
    
   
    if(i==24): #if the pile is going to exhaust and nothing is left to break the tie.
        
        tiebreaker_pile_last_1 =[a[i],a[i+1]]
        tiebreaker_pile_last_2 =[b[i],b[i+1]]
        return(tiebreaker_pile_last_1,inc+4,"Exhausted")
    
    else:      #breaking tie using next cards in player's hand
        
        inc = inc+4
        tiebreaker_pile = []
        tiebreaker_pile.extend([a[i],a[i+1],b[i],b[i+1]])

        for j in range(i+2,25):

            if((a[j]+a[j+1])==(b[j]+b[j+1])):
                 return tiebreaker(a,b,j,inc) #if there is tie again after a tie

            if((a[j]+a[j+1])>(b[j]+b[j+1])): #if Player1 wins in a tiebreaker

                tiebreaker_pile.extend([a[j],a[j+1],b[j],b[j+1]])
                return(tiebreaker_pile,inc,"Player 1")

            elif((a[j]+a[j+1])<(b[j]+b[j+1])): #if player2 wins a tiebreaker.    
        
                tiebreaker_pile.extend([a[j],a[j+1],b[j],b[j+1]])
                return(tiebreaker_pile,inc,"Player 2")    

## Step 3.3

In [893]:
tot_winner,tot_loser,winner,loser = play()
if winner==loser:
    print("There is a tie!")

else:   
    print("Winner is " + winner + " with total of " + str(tot_winner))
    print("Loser is " + loser + " with total of " + str(tot_loser))
    
print("Game terminates after maximum of 13 rounds if there is no tie. If there is one tie, there are 12 rounds, and so on.")    

Player1 wins

Tie first, but Player2 won

Player1 wins

Player1 wins

Player1 wins

Player1 wins

Player2 wins

Player1 wins

Player2 wins

Player2 wins

Player2 wins

Player2 wins

Winner is Player 2 with total of 186
Loser is Player 1 with total of 178
Game terminates after maximum of 13 rounds if there is no tie. If there is one tie, there are 12 rounds, and so on.


# Exercise 4

## Set Up:

In [900]:
def play_analysis():    
    a,b=deal()
   
    initial_a = sum(a)
    initial_b = sum(b)
    
    discard_pile_p1 = []
    discard_pile_p2 = []
    
    i=0

    while i<=25:

        if(i%2==0):

            if((a[i]+a[i+1])==(b[i]+b[i+1])):
                
                tie_pile = []; inc = 0; player_won = ""; p1_exhausted = []; p2_exhausted = [];
                tie_pile,inc,player_won = tiebreaker(a,b,i,inc)
                
                if player_won == "Player 1":
                    discard_pile_p1.extend(tie_pile)

                elif player_won == "Player 2":
                    discard_pile_p2.extend(tie_pile)
                    
                elif player_won == "Exhausted": 
                    
                    p1_exhausted = [a[i],a[i+1]]
                    discard_pile_p1.extend(p1_exhausted)
                    
                    p2_exhausted = [b[i],b[i+1]]
                    discard_pile_p2.extend(p2_exhausted)
                    
                i += inc - 1
                
                
            elif((a[i]+a[i+1])>(b[i]+b[i+1])):
                discard_pile_p1.extend([a[i],a[i+1],b[i],b[i+1]])
                

            elif((a[i]+a[i+1])<(b[i]+b[i+1])):
                discard_pile_p2.extend([a[i],a[i+1],b[i],b[i+1]])
                
        i += 1    
    
    count = 0
    empty = 0
    winner_initial = []
    loser_initial = []
    
    #print("Will return now")
    #print(initial_a)
    #print(initial_b)
    #print(sum(discard_pile_p1))
    #print(sum(discard_pile_p2))
    #print("")
    
    if((sum(discard_pile_p1)>sum(discard_pile_p2))):  #Taking condition where player1's pile is greater
        
        if(initial_a<initial_b):  #Taking condition where player1 won instead of low initial total
            
            count = 1
            return sum(discard_pile_p1),sum(discard_pile_p2),count,initial_a,initial_b,initial_a,initial_b
            
        elif((initial_a>initial_b) or (initial_a==initial_b)): #Taking condition where player1 won after high or equal initial value
            return sum(discard_pile_p1),sum(discard_pile_p2),count,empty,empty,initial_a,initial_b
            

    elif((sum(discard_pile_p1)<sum(discard_pile_p2))): #Taking condition where player2's pile is greater
        
        if(initial_b<initial_a): #Taking condition where player2 won instead of low initial total
            
            count = 1
            return sum(discard_pile_p1),sum(discard_pile_p2),count,initial_b,initial_a,initial_b,initial_a
        
        elif((initial_b>initial_a) or (initial_b==initial_a)): #Taking condition where player1 won after high or equal initial value
            return sum(discard_pile_p1),sum(discard_pile_p2),count,empty,empty,initial_b,initial_a
        
    elif(sum(discard_pile_p1)==sum(discard_pile_p2)): # Taking condition where there is tie between player 1 and player 2 after all rounds
        #This is considered as 0 when we take situation of low initial value but won.
        print("Output of discard pile is equal, game is tie")
        return sum(discard_pile_p1),sum(discard_pile_p2),count,empty,empty,initial_a,initial_b    
        
    
def tiebreaker(a,b,i,inc):
    
   
    if(i==24):
        
        tiebreaker_pile_last_1 =[a[i],a[i+1]]
        tiebreaker_pile_last_2 =[b[i],b[i+1]]
        return(tiebreaker_pile_last_1,inc+4,"Exhausted")
    
    else:
        
        inc = inc+4
        tiebreaker_pile = []
        tiebreaker_pile.extend([a[i],a[i+1],b[i],b[i+1]])

        for j in range(i+2,25):

            if((a[j]+a[j+1])==(b[j]+b[j+1])):
                 return tiebreaker(a,b,j,inc)

            if((a[j]+a[j+1])>(b[j]+b[j+1])):

                tiebreaker_pile.extend([a[j],a[j+1],b[j],b[j+1]])
                return(tiebreaker_pile,inc,"Player 1")

            elif((a[j]+a[j+1])<(b[j]+b[j+1])):    
        
                tiebreaker_pile.extend([a[j],a[j+1],b[j],b[j+1]])
                return(tiebreaker_pile,inc,"Player 2")    

### 20 simulation using function call to play_analysis function in for loop and getting the returned values in appropriate dictionaries.

In [902]:
import numpy as np

tot_winnera = {}; tot_losera = {}; count = {}; low_win = {}; high_loss = {}; winner_initial = {}; loser_initial = {};

for i in range(0,20,1):
    tot_winnera[i],tot_losera[i],count[i], low_win[i], high_loss[i],winner_initial[i],loser_initial[i] = play_analysis()

## Step 4.1

In [903]:
for i in range(0,20,1):
    print("Game " + str(i+1))
    print("Winner total: " + str(tot_winnera[i]))
    print("Loser total: " +str(tot_losera[i]))
    print("")
    
print("Maximum of all winning values:")
print(max(tot_winnera.values()))

print("Minimum of all winning values:")
print(min(tot_winnera.values()))

print("Maximum of all losing values:")
print(max(tot_losera.values()))

print("Minimum of all losing values:")
print(min(tot_losera.values()))

print("Mean of all winning values:")
print((sum(tot_winnera.values())/len(tot_winnera)))

print("Mean of all losing values:")
print((sum(tot_losera.values())/len(tot_losera)))    

Game 1
Winner total: 156
Loser total: 208

Game 2
Winner total: 158
Loser total: 206

Game 3
Winner total: 173
Loser total: 191

Game 4
Winner total: 177
Loser total: 187

Game 5
Winner total: 181
Loser total: 183

Game 6
Winner total: 297
Loser total: 67

Game 7
Winner total: 224
Loser total: 140

Game 8
Winner total: 143
Loser total: 221

Game 9
Winner total: 166
Loser total: 198

Game 10
Winner total: 203
Loser total: 161

Game 11
Winner total: 159
Loser total: 205

Game 12
Winner total: 151
Loser total: 213

Game 13
Winner total: 337
Loser total: 27

Game 14
Winner total: 176
Loser total: 188

Game 15
Winner total: 257
Loser total: 107

Game 16
Winner total: 204
Loser total: 160

Game 17
Winner total: 231
Loser total: 133

Game 18
Winner total: 228
Loser total: 136

Game 19
Winner total: 239
Loser total: 125

Game 20
Winner total: 184
Loser total: 180

Maximum of all winning values:
337
Minimum of all winning values:
143
Maximum of all losing values:
221
Minimum of all losing value

#### Addition of winning and losing total is always 364. If total is around 202, you can win, considering mean for winning total or if value is about 160, you can lose, considering mean of losing total.

## Step 4.2

In [904]:
low_tot_win = np.array(list(low_win.values()))

print("Proportion of game that resulted in winner that had lower initial total:")
print((sum(count.values())/len(count)))

if(np.count_nonzero(low_tot_win)==0):
    print("No situation where initial lower value wins therefore there is no lowest initial total to win a game or conversely, the highest initial total to lose a game")
else:
    lowest_initial_win=np.min(low_tot_win[np.nonzero(low_tot_win)])
    print("Lowest initial total to win:")
    print(lowest_initial_win)
    print("Highest initial total to lose:")
    print(max(high_loss.values()))
    

Proportion of game that resulted in winner that had lower initial total:
0.3
Lowest initial total to win:
168
Highest initial total to lose:
196


## Step 4.3

In [905]:
arr_final_winner =  np.array(list(tot_winnera.values()))
#print(np.shape(arr_final_winner))
arr_final_loser =  np.array(list(tot_losera.values()))
#print(arr_final_loser)
arr_initial_winner =  np.array(list(winner_initial.values()))
#print(arr_initial_winner)
arr_initial_loser =  np.array(list(loser_initial.values()))
#print(arr_initial_loser)
    
print("Correlation between final winning total and initial winning total:")    
print(np.corrcoef(arr_initial_winner,arr_final_winner,rowvar = True))
print("Correlation between final losing total and initial losing total:")
print(np.corrcoef(arr_initial_loser,arr_final_loser, rowvar = True))

Correlation between final winning total and initial winning total:
[[1.         0.76295383]
 [0.76295383 1.        ]]
Correlation between final losing total and initial losing total:
[[1.         0.76295383]
 [0.76295383 1.        ]]


The (linear) correlation between the initial and final totals for winners and losers is positive and equal(0.76), which tells that there is +ve relationship between initial and final, that is direct proportion which states that , usually a higher initial total will give a higher final total and a lower initial total will give a lower final total.

Also, sometimes a realtion can show negative value, only when the condition about lower initial total still a higher final total is considered as we got proportion for some in above steps, but it is a rare case.

The conclusion we can take from this is for the simulation of 20 games, the positive value is showing a strong direct proportion which tells us that if we take more number of simulations or more values to find the result, the correlation will increse implying a strong possibility of getting high final score if there is high initial score.